storage/indexer/src/event_v2_translator.rs

```diff
@@ -0,0 +1,1327 @@
+        translators.insert(CANCEL_OFFER_TYPE.clone(), Box::new(CancelOfferTranslator));
+        translators.insert(CLAIM_TYPE.clone(), Box::new(ClaimTranslator));
+        translators.insert(
+            COLLECTION_DESCRIPTION_MUTATE_TYPE.clone(),
+            Box::new(CollectionDescriptionMutateTranslator),
+        );
+        translators.insert(
+            COLLECTION_URI_MUTATE_TYPE.clone(),
+            Box::new(CollectionUriMutateTranslator),
+        );
+        translators.insert(
+            COLLECTION_MAXIMUM_MUTATE_TYPE.clone(),
+            Box::new(CollectionMaximumMutateTranslator),
+        );
+        translators.insert(URI_MUTATION_TYPE.clone(), Box::new(UriMutationTranslator));
+        translators.insert(
+            DEFAULT_PROPERTY_MUTATE_TYPE.clone(),
+            Box::new(DefaultPropertyMutateTranslator),
+        );
+        translators.insert(
+            DESCRIPTION_MUTATE_TYPE.clone(),
+            Box::new(DescriptionMutateTranslator),
+        );
+        translators.insert(
+            ROYALTY_MUTATE_TYPE.clone(),
+            Box::new(RoyaltyMutateTranslator),
+        );
+        translators.insert(
+            MAXIMUM_MUTATE_TYPE.clone(),
+            Box::new(MaximumMutateTranslator),
+        );
+        translators.insert(
+            OPT_IN_TRANSFER_TYPE.clone(),
+            Box::new(OptInTransferTranslator),
+        );
+        Self {
+            main_db_reader,
+            internal_indexer_db,
+            translators,
+            event_sequence_number_cache: DashMap::new(),
+        }
+    }
+
+    // When the node starts with a non-empty EventSequenceNumberSchema table, the in-memory cache
+    // `event_sequence_number_cache` is empty. In the future, we decide to backup and restore the
+    // event sequence number data to support fast sync, we may need to load the cache from the DB
+    // when the node starts using this function `load_cache_from_db`.
+    pub fn load_cache_from_db(&self) -> Result<()> {
+        let mut iter = self
+            .internal_indexer_db
+            .iter::<EventSequenceNumberSchema>()?;
+        iter.seek_to_first();
+        while let Some((event_key, sequence_number)) = iter.next().transpose()? {
+            self.event_sequence_number_cache
+                .insert(event_key, sequence_number);
+        }
+        Ok(())
+    }
+
+    pub fn cache_sequence_number(&self, event_key: &EventKey, sequence_number: u64) {
+        self.event_sequence_number_cache
+            .insert(*event_key, sequence_number);
+    }
+
+    pub fn get_cached_sequence_number(&self, event_key: &EventKey) -> Option<u64> {
+        self.event_sequence_number_cache
+            .get(event_key)
+            .map(|seq| *seq)
+    }
+
+    pub fn get_next_sequence_number(&self, event_key: &EventKey, default: u64) -> Result<u64> {
+        if let Some(seq) = self.get_cached_sequence_number(event_key) {
+            Ok(seq + 1)
+        } else {
+            let seq = self
```

Comment:
in what case you will find the data in database but not in cache?

Reply:
The cache and the database don't need to be perfectly synced as we don't plan to support the fast sync at the moment. When a node stops and resumes, we dont' fill the cache with the data in the DB, and so the data may be present in the DB but not in the cache.

Reply:
If we fully load it on startup, memory usage would continuously increase even after restarting the node. If we decide to support fast sync in the future, fully loading at startup would become necessary. At that point, I can revisit and update the code to accommodate that requirement.